### DISCUSSION TITLE
Can I use `coreutils` as a library?

### DISCUSSION THREAD

**Discussion by liblaf** [uutils/coreutils#4408] [Q&A] [answered]
As described in title, can I use this crate as a library?
I want to call utils in coreutils while avoid using `Command`.

**Comment by tertsdiepraam** [comment, marked as answer]
Technically somewhat yes, but in practice no.

Each of the utils is published as a crate and you can call their `main` function with some arguments. However, it still behaves exactly as the command line utils, meaning that:

1. The arguments you give are a list of strings that are parsed like on the command line.
2. It only returns the exit code all the other output is printed to stdout.

So the utils are not very useful as a library, but if the above is exactly what you need, it's possible. This is how the multicall binary works under the hood.

> Note: we make no guarantees about the stability of using them as a library.

Does that answer your question?

**Reply by liblaf** [reply]
THX, that's good enough for me!

**Reply by tertsdiepraam** [reply]
Oh I forgot to mention an important aspect. The utils might exit on their own and not return. The style we write in nowadays usually returns, but there are definitely some `std::process::exit` calls. Even failed argument parsing could just exit the program.

**Reply by liblaf** [reply]
> Oh I forgot to mention an important aspect. The utils might exit on their own and not return. The style we write in nowadays usually returns, but there are definitely some `std::process::exit` calls. Even failed argument parsing could just exit the program.

emmmm, so I have to use something like `fork()` to prevent the program from exit? I haven't seen a higher-level interface to directly call a function from a new process / subprocess.

**Reply by sylvestre** [reply]
Some of the core functions of uucore might be easily usable. https://uutils.github.io/dev/uucore/index.html
